Backlash cleanup. Save backlash, fil. sensor, ExtUI userdata to EEPROM

marcio-ao opened the pull request:
- Consolidated all the backlash related code into its own ".h" and ".cpp" file
- Added backlash settings to EEPROM (backlash_correction, backlash_distance_mm[XYZ], backlash_smoothing_mm = 17 bytes)
- Added filament runout settings to EEPROM (runout_sensor_enabled + runout_distance_mm = 5 bytes)
- Added 32-bytes [1] of user storage for ExtUI.

@thinkyhead:  My understanding on how the EEPROM macros work is limited, please look this over and make sure I am doing things correctly.

[1] The size of 32 bytes was chosen because this is about how much the Lulzbot touch UI will require to be made compatible with all boards. This storage will be used for sound volume, screen brightness, touch point calibration matrix, sound scheme preferences and screen lock password. It is anticipated that each ExtUI will have different needs, but 32 bytes seems like a reasonable starting point.

marcio-ao committed:
Backlash updates and improvements

thinkyhead commented:
Settings stored in EEPROM need G-code backing, so for `runout_distance_mm` we ought to add `M412 D` and include this in the EEPROM output for `M503`.

marcio-ao commented:
@thinkyhead: I have a question about the EEPROM code. It seems like there are two possibilities for dealing with options that are enabled or disabled:

1) The layout of the EEPROM is the same regardless of which options are enabled. This would mean that if an option is disabled at compile-tine, padding values are still written to the EEPROM
2) The layout of the EEPROM varies depending on which options are enabled.

Looking at "configuration_store.cpp", some things suggest that 1) is the goal, while other things make it seems like 2) is the goal.

For example, if 1) is the goal, then the struct SettingsData should not have any conditional statements. But yet it does. The size of the SettingsData structure varies based on `HAS_HOTEND_OFFSET`, `ENABLED(MESH_BED_LEVELING)`, `GRID_MAX_POINTS_X`, `GRID_MAX_POINTS_Y`, `AUTO_BED_LEVELING_BILINEAR`, `ENABLED(DELTA)`, `EXTUDERS > 1`

So that suggests that in practice, 2) is true. If that is the case, all the conditional later on in the file which write padding bytes does not make sense.

So, I'm at a loss here. I don't really understand what the goal is.

marcio-ao commented:
@thinkyhead: Also, it seems like the _FIELD_TEST() stuff might want to make use of static_asserts, although I do not know if that feature is getting held back for now for compatibility reasons.

thinkyhead commented:
The goal has been to keep the EEPROM structure the same if the machine architecture is the same. So on a Cartesian enabling or disabling options should ideally leave the structure unchanged. Exceptions exist. Adding an EXTRUDER alters the structure.

Not every contributor has followed this maxim, and it is not always very easy to follow.. So there may be holes. Fill them in when you find them.

I don’t think FIELD TEST can be made static constexpr. It compares a variable.

thinkyhead commented:
I plan to rewrite the EEPROM code soon. It will use IFF format instead of being a blind binary blob.

marcio-ao commented:
> The goal has been to keep the EEPROM structure the same if the machine architecture is the same. So on a Cartesian enabling or disabling options should ideally leave the structure unchanged.

@thinkhead: I see. So in the case of the stuff I am adding, backlash and runout sensor variables should reserve space.

How about the ExtUI? Would that be part of the machine architecture, or part of the options? That's kind of a gray area.

> I don’t think FIELD TEST can be made static. It compares a variable.

Gotcha. I missed that. There is a index variable. Sometimes I really wish C++ had the concept of a compile-time variable that could be incremented, but would otherwise behave like a constexpr.

Although why not simply create a SettingsDataStruct on the stack, copy all the values into it, then write the whole data block to EEPROM with a loop? Rather than having to verify that data is written in the order that corresponds to SettingsDataStruct, just write SettingsDataStruct itself.

marcio-ao commented:
> I plan to rewrite the EEPROM code soon. It will use IFF format instead of being a blind binary blob.

It seems like IFF requires tags and lengths for the various chunks. Given that EEPROM is only 1024 bytes, wouldn't it be wasteful even if the tags were just a byte? I don't think the current scheme is bad per se...

marcio-ao commented:
> The goal has been to keep the EEPROM structure the same if the machine architecture is the same. So on a Cartesian enabling or disabling options should ideally leave the structure unchanged.

@thinkyhead: Maybe "machine architecture" should be interpreted as "things that require hardware changes". So in that case, backlash compensation would be an option that reserves space in the EEPROM, but filament sensor and ExtUI would be hardware changes that wouldn't. I'll make that adjustment.

thinkyhead commented:
> Given that EEPROM is only 1024 bytes, wouldn't it be wasteful even if the tags were just a byte?

We have to "waste" some space if we want structured data that can be interpreted in a forward- and bacward-compatible way. It's likely that a good amount of space will be saved by not having placeholders for disabled features. It is worth developing, and then we will see whether it fills up 1024 bytes more quickly than the current schema.

thinkyhead commented:
> filament sensor and ExtUI would be hardware changes that wouldn't

Under the current scheme, *any feature* that would add data to the EEPROM should get a placeholder when disabled.

thinkyhead commented:
> ExtUI?

We can consider changing the screen a big enough change that we don't need placeholding for that.

marcio-ao commented:
> Under the current scheme, any feature that would add data to the EEPROM should get a placeholder when disabled.

Do you want me to change the filament sensor back to reserving space as it was before? It makes more sense to me for it not to reserve space, because a printer either has or does not have a filament sensor. Having placeholders is a convenience to people who may be trying different software settings, but if a change already requires adding or removing hardware, or rewiring their board, it doesn't seem like clearing the EEPROM and starting from scratch is that inconvenient. Plus it ultimately saves on EEPROM space, because not every printer is going to have every hardware feature that is available.

> We can consider changing the screen a big enough change that we don't need placeholding for that.

Sounds good. Once you move to using IFF, we can consider giving the ExtUI a unique tag, or possibly giving each ExtUI display type a unique tag.

thinkyhead commented:
> Do you want me to change the filament sensor back to reserving space as it was before?

It's up to your discretion. If the feature is small and only saves 10 bytes then it's no big deal to retain the placeholder, and it will not raise alarms if someone notices the inconsistency w/r/t other settings.

thinkyhead commented:
> possibly giving each ExtUI display type a unique tag

The more generic the EEPROM storage is, the better. And remember, everything in EEPROM requires a G-code so that settings can be reported/saved/restored in the form of G-code.

marcio-ao commented:
> It's up to your discretion. If the feature is small and only saves 10 bytes then it's no big deal to retain the placeholder, and it will not raise alarms if someone notices the inconsistency w/r/t other settings.

@thinkyhead: I was the one who noticed the inconsistency and I feel it would have made sense if there had been a code comment explaining the rationale. I added some comments. Let me know if it is you're okay with the things how they are now.

thinkyhead commented:
In spite of my statement of justification, there's no hard and fast rule about these settings related to "hardware changes" per se, but we have been aiming to try to keep the EEPROM structure consistent so it doesn't break so easily on common config changes. Some types of hardware add-ons are so common that it is sensible to have a placeholder for it. Filament sensors would be a classic example.

marcio-ao commented:
> Some types of hardware add-ons are so common that it is sensible to have a placeholder for it. Filament sensors would be a classic example.

@thinkyhead: Okay. I'll make the change then.

thinkyhead commented on Marlin/src/feature/runout.cpp:
```diff
@@ -51,7 +51,7 @@ void FilamentSensorBase::filament_present(const uint8_t extruder) {
 #endif
 
-#if FILAMENT_RUNOUT_DISTANCE_MM > 0
+#ifdef FILAMENT_RUNOUT_DISTANCE_MM
```
Allowing for a negative or zero `FILAMENT_RUNOUT_DISTANCE_MM`? If not, a sanity check will be needed to support this change.

marcio-ao replied:
Definitely zero. The idea is a vendor like ourselves can set this to zero, to enable the code, while still allowing the user to change it using M425 or LCD.

For negative, no, that is not expected.

thinkyhead commented on Marlin/src/gcode/feature/runout/M412.cpp:
```diff
@@ -31,22 +31,28 @@
     if (seenS) runout.enabled = parser.value_bool();
+    #ifdef FILAMENT_RUNOUT_DISTANCE_MM
+      if (seenD) RunoutResponseDelayed::runout_distance_mm = parser.value_linear_units();
+    #endif
```
Assigning `seenD` before reading the value for `S` breaks this! You must call `parser.value...` immediately after `seen` for the parameter word whose value you want. Since `S` is last, the correct value will be queued up for `S`.

thinkyhead replied:
Fixed by moving `seen('D')` into the `if` clause.

thinkyhead committed:
backlash.backlash_ => backlash.

thinkyhead commented on Marlin/src/module/configuration_store.cpp:
```diff
@@ -1895,18 +1895,18 @@ void MarlinSettings::postprocess() {
         #if ENABLED(BACKLASH_COMPENSATION)
-          float   *backlash_distance_mm      = backlash.backlash_distance_mm;
-          uint8_t &backlash_correction       = backlash.backlash_correction;
+          float   (&backlash_distance_mm)[XYZ] = backlash.distance_mm;
```
This is how you make a reference that knows the size of the array it points to, so `sizeof()` will return 12.

marcio-ao replied:
Oh, okay. Good to know.

thinkyhead committed:
Update M412.cpp

marcio-ao committed:
Added "planner.h" for block_t.

marcio-ao committed:
Fix compile error

When DISABLED(BACKLASH_COMPENSATION), BACKLASH_CORRECTION will be undefined.

marcio-ao committed:
Reworked how ExtUI persistent data is handled.

- Now the ExtUI is called during an EEPROM read or write to copy data in and out of a buffer.
- This prevents global memory from needing to be reserved.
- Also simplifies ExtUI implementation by not forcing the ExtUI to keep data in a specific location.

marcio-ao committed:
Commented out example bits that won't compile.

marcio-ao committed:
Merge branch 'bugfix-2.0.x' into pr-backlash-cleanup-and-new-eeprom-settings

marcio-ao committed:
Moved initialization of UI before EEPROM read.

- This change is necessary for the ExtUI, does not appear to affect UltraLCD

- Reading from the EEPROM causes ExtUI related settings to be loaded.
- This initiates a call to ExtUI::loadSettings()
- Without this change, the ExtUI would be called before it had initialized.
- This change ensures that "ui.init()" is called before any other ui method.

marcio-ao committed:
Merge branch 'pr-backlash-cleanup-and-new-eeprom-settings' of github.com:marcio-ao/Marlin into pr-backlash-cleanup-and-new-eeprom-settings

marcio-ao committed:
Increased storage space to 48 bytes.

marcio-ao committed:
Check that STATS_EEPROM_ADDRESS does not overlap

marcio-ao committed:
Merge branch 'bugfix-2.0.x' into pr-backlash-cleanup-and-new-eeprom-settings

marcio-ao committed:
Fixed duplication caused by conflicting PR.

thinkyhead committed:
Update Marlin.cpp

thinkyhead committed:
DO NOT wrap whole headers in config conditions

Instead, just don't include the header when the feature is not enabled.

Since is often necessary to use configured values in a header, of course it's appropriate to include `MarlinConfigPre.h` and use `ENABLED`, etc.

marcio-ao committed:
Added missing include file.

thinkyhead committed:
Update configuration_store.cpp

thinkyhead committed:
Update configuration_store.cpp

marcio-ao committed:
Fix compilation error.

marcio-ao committed:
Merge branch 'pr-backlash-cleanup-and-new-eeprom-settings' of github.com:marcio-ao/Marlin into pr-backlash-cleanup-and-new-eeprom-settings

marcio-ao committed:
Trying to fix compilation error.

marcio-ao commented:
@thinkyhead: It looks like travis is borked at an "apt-get install"

```
E: Unable to locate package g++-7
E: Couldn't find any package by regex 'g++-7'
```

p3p commented:
poked it with a stick, appears to be running again.

marcio-ao commented:
> poked it with a stick, appears to be running again.

Travis is now blind in one eye.

marcio-ao commented:
Oooooo. It's green again. Excellent! This PR has been more in the red than in the green...

thinkyhead commented:
We love working for the green checkmarks, even if sometimes it is just fighting Travis implosions.

thinkyhead committed:
Clean up some formatting

thinkyhead committed:
Add accessors for runout distance

thinkyhead commented:
Ok, this seems pretty solid now! Just waiting for one more pass, and then it will be merged.

marcio-ao commented:
Travis is temperamental today.

p3p commented:
That would be easier if PlatformIO hadn't just decided to break in new and exciting ways, even running the tests (or building) locally is failing

```*** [.pioenvs/LPC1768/firmware.elf] Implicit dependency `/home/p3p/.platformio/platforms/nxplpc-arduino-lpc176x/builder/${common.build_flags}' not found, needed by target `.pioenvs/LPC1768/firmware.elf'.```

p3p commented:
Probably something to do with `PlatoformIO 4.0.0a9` and `Include external configuration files with "extra_configs"`, why we use the dev branch for the platformio install in Travis is an unrelated issue ;)

thinkyhead committed:
Malyan refactor as ExtUI

thinkyhead commented:
Don't bring Plato into this. Aristotle is clearly the more at fault.

marcio-ao commented:
What's a Malyan display? The one that comes up when I search looks like a [clone of the reprapdiscount display](https://hobbyking.com/en_us/malyan-m150-i3-3d-printer-replacement-lcd-screen.html?countrycode=US&gclid=EAIaIQobChMIpKy63aSA4gIVmONkCh3UiQlzEAQYASABEgJh2PD_BwE&gclsrc=aw.ds)

thinkyhead commented:
It comes with the Malyan M200, apparently.
